torch/distributed/nn/_functions.py

```diff
@@ -0,0 +1,339 @@
+        # or a different device for this rank and call the appropiate
+        # function
+        tensors = [t.clone() for t in tensors]
+        dist.broadcast_multigpu(tensors, src, group=group, src_tensor=0)
+        return tuple(tensors)
+
+    @staticmethod
+    def backward(ctx, *grad_outputs):
+        # Number of returns change depending on if we used multigpu or not
+        return (None, None, None) + _ReduceMultiGPU.apply(ctx.src, dist.ReduceOp.SUM, ctx.group, ctx.src_tensor, *grad_outputs)
+
+
+class _Gather(Function):
+    @staticmethod
+    def forward(ctx, src, group, tensor):
+        ctx.src = src
+        ctx.group = group
+        # Need to create a list of tensors here to do the
+        # aggregation, get it from the group size
+        # tensor should be correctly sized for the method
+        # gathering
+        tensor_list = [
+            torch.empty_like(tensor) for i in range(dist.get_world_size())
+        ]
+        ctx.rank = dist.get_rank()
+        dist.gather(tensor, tensor_list, src, group=group)
+        return tuple(tensor_list)
+
+    @staticmethod
+    def backward(ctx, *grad_outputs):
+        return (None, None) + (_Scatter.apply(ctx.src, ctx.group, *grad_outputs),)
+
+
+class _Scatter(Function):
+    @staticmethod
+    def forward(ctx, src, group, *tensors):
+        ctx.src = src
+        ctx.group = group
+        # TODO assert all the tensors have the same size
+        output = torch.empty_like(tensors[0])
+        dist.scatter(output, tensors, src, group=group)
+        return output
+
+    @staticmethod
+    def backward(ctx, *grad_outputs):
+        return (None, None) + _Gather.apply(ctx.src, ctx.group, *grad_outputs)
+
+
+class _Reduce(Function):
+    @staticmethod
+    def forward(ctx, src, op, group, tensor):
+        ctx.src = src
+        ctx.group = group
+        dist.reduce(tensor, src, op=op, group=group)
+        return tensor
+
+    @staticmethod
+    def backward(ctx, *grad_outputs):
+        return (None, None, None) + (_Broadcast.apply(ctx.src, ctx.group, *grad_outputs),)
+
+
+class _ReduceMultiGPU(Function):
+    @staticmethod
+    def forward(ctx, dst, op, group, dst_tensor, *tensors):
+        ctx.dst = dst
+        ctx.group = group
+        dist.reduce_multigpu(tensors, dst, op=op, group=group)
+        return tuple(tensors)
+
+    @staticmethod
+    def backward(ctx, *grad_outputs):
+        return (None, None, None, None) + _BroadcastMultiGPU.apply(ctx.src, ctx.group, *grad_outputs)
+
+
+
+class _AllGather(Function):
+    @staticmethod
+    def forward(ctx, group, tensor):
+        ctx.group = group
+        out_tensor_list = [
+            torch.empty_like(tensor) for i in range(dist.get_world_size())
+        ]
+        dist.all_gather(out_tensor_list, tensor, group=group)
+        return tuple(out_tensor_list)
+
+    @staticmethod
+    def backward(ctx, *grad_outputs):
+        gxs = _AlltoAll.apply(ctx.group, *grad_outputs)
+        gx = torch.stack(gxs).sum(dim=0)
```

Comment:
nit: you can do the same think with `gx = sum(gxs)` this will avoid the extra allocation.